### PR TITLE
Dist removal

### PR DESCRIPTION
**Pull request by mbargull**
### Description

Work in progress to remove legacy `conda.models.dist.Dist` usage.
Excerpt from gh-5069 :
> \[...\] `conda-build`'s slow (and memory intensive) ramp up before it gets to its actual first environment solving.
> \[...\] work on purging legacy behavior/code such that we could at least prepare to make use of the deferred creation of package records.
> A first step in that direction would be to finally remove `conda-build`'s usage of the old `Dist` objects (removed from `conda`'s core workings in 2018).
> 
> ----
> #### Impact
> - Prepare usage of newer/more performant package index handling in `conda-build`.
> - Removal and possible removal tech debt not only in `conda-build` but also `conda`.
> - **Breaking** changes for library uses of `conda_build`:
>     - not only might we remove `Dist`-related functions which might be used downstream,
>     - but also not using a `Dist`-based package index anymore could break downstream expectations.

Closes gh-5069
### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by mbargull** [comment]
Pull request overiew:
#### Goal
- Change necessary parts to remove the usage of `Dist`.
- Do not change how environment creations happen just yet.
    => This means: Actual runtime benefits (time/memory reduction) should be done *after* the `Dist` removal/cleanup work. (Ideally in subsequent PRs.)

#### Strategy
- Do not change behavior.
- Work iteratively and surgically.
- Copy over `conda.models.dist` and `conda.plan` (+`conda.instructions`).
    (These still exist (nearly) completely for `conda-build` only and could otherwise be removed from `conda`.)
- Iteratively reduce the surface of `Dist`/`plan` interaction.
    - remove `Dist` usage and unneeded functionality from `Dist` and `plan`
        (e.g., `conda-build` only creates, never updates environments => only create/link parts of `plan` are actually used)

#### Progress
- `Dist` objects have been removed.
- Test suite not yet run (hence opening this draft PR).
    - (Expecting failures; only ran local `conda-forge/zlib-feedstock` builds not yet more of the test suite, yet.)
- Removed most unneeded/defunct code regarding `plan`.
- More cleanup needed.

**Comment by mbargull** [comment]
The changes here are now ready for review.
Overall, this just removes `Dist` and related functionality but not change behavior.
Changes in behavior (e.g., replacing the index and solver invocation in `install_actions`) should go into a subsequent PR.

If I get to it in the next days, I'd like to clean up the commit messages a bit.
I'll also add some comments later here.

**Comment by jezdez** [comment]
@mbargull Since this is a 24.1.0 related refactor, I'm inclined to review now as I'm starting to prepare the release. Do you feel comfortable with including it in 24.1.0?

**Comment by mbargull** [comment]
@jezdez, yes, since this mostly refactoring / code removal, I think we can include this in the upcoming release.
We can discuss further (post-`24.1`) action later today in the build tools meeting.
I'll add news entry etc later today (currently afk for a few hours).

**Comment by mbargull** [comment]
I'll re-review this in a couple of hours and probably re-add some of the removed functions/symbols.

In the build tools meeting we had raised these points:
1. We probably want to include this in the `24.1` release, if possible.
2. Although this has a lot of incremental commits, the overall changeset became relatively small -- very much thanks to @kenodegard's work on cleaning up/deprecating things which heavily relied on `Dist` things beforehand!
3. Even though this doesn't use some removed (imported) functions now, we probably want to still (re-)export those for now because some are used in conda-forge's infrastructure-related scripts, `boa` etc., mentioned by @jaimergp in November's meeting:
    - https://github.com/search?q=org%3Aconda-forge+%2F%28from%7Cimport%29+conda_build%2F+language%3APython&type=code
    - https://github.com/search?q=org%3Amamba-org+%2F%28from%7Cimport%29+conda_build%2F+language%3APython&type=code

    We should mark these for deprecation then, of course (same for `conda.models.dist`/`conda.plan`).
4. I have to open an issue for replacing `conda.plan`'s (with this PR now in `conda_build.environ`) functionality.
    (We may want to mark some additional things for deprecation in `24.1` so we can do necessary removals in a future version.)

File: conda_build/conda_interface.py
Comment (kenodegard) [inline conda_build/conda_interface.py:97]: `deprecated.constant` uses the module level `__getattr__` to detect usage of deprecated variables which only occurs if the variable doesn't exist, we need to change all of these imports into:

```python
from conda.models.dist import Dist as _Dist
```

```suggestion
deprecated.constant("24.1.0", "24.3.0", "Dist", _Dist)
deprecated.constant("24.1.0", "24.3.0", "display_actions", _display_actions)
deprecated.constant("24.1.0", "24.3.0", "execute_actions", _execute_actions)
deprecated.constant("24.1.0", "24.3.0", "execute_plan", _execute_plan)
deprecated.constant("24.1.0", "24.3.0", "get_index", _get_index)
deprecated.constant("24.1.0", "24.3.0", "install_actions", _install_actions)
deprecated.constant("24.1.0", "24.3.0", "linked", _linked)
deprecated.constant("24.1.0", "24.3.0", "linked_data", _linked_data)
```

**Review by kenodegard — COMMENTED** [review]
FYI I just opened https://github.com/conda/conda-build/pull/5151 to remove all 24.1.0 deprecations

> If I get to it in the next days, I'd like to clean up the commit messages a bit.

Don't worry about too much about the commit messages, we only do squash commits.

File: conda_build/environ.py
Comment (mbargull) [reply conda_build/environ.py]: I now used the `tuple(dict.fromkeys(...))` (i.e., leveraging `dict` preserving input order for now and therefore using it's keys as an order-preserving set) pattern so we can avoid the extra dependency.

File: conda_build/conda_interface.py
Comment (mbargull) [reply conda_build/conda_interface.py]: Good call, thanks.

File: conda_build/utils.py
Comment (mbargull) [reply conda_build/utils.py]: > It uses `str.rstrip` is incase the build string is empty?

Yes, just for those unlikely cases.
Technically, the code this replaces didn't have the `rstrip()`, so we could remove it here.
It just felt odd to return a requirement string with trailing whitespace.
However, I think we may want to keep it as before without `rstrip` for now and added a commit to do so.
Let's change these subtleties whenever we might refactor/cleanup code around its usage.

**Comment by mbargull** [comment]
> FYI I just opened #5151 to remove all 24.1.0 deprecations

Thanks, I'll merge this gh-5151 here once it's in.

**Review by jezdez — DISMISSED** [review]

File: conda_build/conda_interface.py
Comment (kenodegard) [inline conda_build/conda_interface.py]: Whoops, accidentally added back the `IndexRecord`:

```suggestion

from .deprecations import deprecated

```

File: conda_build/environ.py
Comment (kenodegard) [inline conda_build/environ.py:22]:
```suggestion
from conda.base.constants import DEFAULTS_CHANNEL_NAME, UNKNOWN_CHANNEL
```

File: CHANGELOG.md
Comment (kenodegard) [inline CHANGELOG.md:91]:
```suggestion
* Remove dependency on `conda.plan`. (#5074)
* Remove almost all dependency on `conda.models.dist`. (#5074)
* Replace usage of legacy `conda.models.dist.Dist` with `conda.models.records.PackageRecord`. (#5074)
```

**Review by kenodegard — APPROVED** [review]
looking good! 🚀

**Review by kenodegard — APPROVED** [review]

File: CHANGELOG.md
Comment (mbargull) [inline CHANGELOG.md:28]: @kenodegard, I forgot to add the `linked_data_no_multichannels` before and saw that you missed the `IndexRecord` stuff in the merge -- which is why I blatantly force-pushed the merge commit I prepared :open_mouth:

**Comment by mbargull** [comment]
Thanks everyone for reviewing!

**Comment by kenodegard** [comment]
@mbargull phew! thanks soooo much for working on this (and being patient with our slooooow CI)! we would not have gotten this done nearly as quickly if you hadn't helped out with the `Dist` removal

BTW since you have merge rights can you click auto-merge and update the squashed commit message?

**Comment by mbargull** [comment]
> (and being patient with the slooooow CI)

All good :).
I was also surprised about the skeleton test failure and had to dig a bit to understand/remember those permission issues (previously encountered those some years ago with clean up code for `go` packages which marked directories as read-only).
At some point I'd also like to continue to look into the parts in `conda-build` that make it so slow on Windows/macOS so we may get sub 30min CI times on all platforms :crossed_fingers: :).

> BTW since you have merge rights can you click auto-merge and update the squashed commit message?

Oh right, good idea! Done.

**Pull request merged**
